### PR TITLE
For #544. Add Coinbase provider

### LINKED ISSUE #544: Coinbase OAuth Provider Support
**Is your feature request related to a problem? Please describe.**
N/A

**Describe the solution you'd like**
Coinbase offers [an Oauth2 API](https://developers.coinbase.com/api/v2#oauth2-coinbase-connect) and it would be super appreciated to have it integrated with this project.

**Describe alternatives you've considered**
Not really an alternative, but some work around this has been done via @bchavez's great https://github.com/bchavez/Coinbase implementation.

**Additional context**
Thank you for any consideration and all your great work out there. 👍

### PR DESCRIPTION
**Pull request by levimatheri**
Add Coinbase provider

### CODE REVIEW

File: AspNet.Security.OAuth.Coinbase/CoinbaseAuthenticationDefaults.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.AuthorizationEndpoint"/>.
+        /// </summary>
+        public const string AuthorizationEndpoint = "https://www.coinbase.com/oauth/authorize";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.TokenEndpoint"/>.
+        /// </summary>
+        public const string TokenEndpoint = "http://www.coinbase.com/oauth/token";
```
Comment (martincostello) [inline AspNet.Security.OAuth.Coinbase/CoinbaseAuthenticationDefaults.cs]:
```suggestion
        public const string TokenEndpoint = "https://www.coinbase.com/oauth/token";
```

File: test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj [hunk 2]
```diff
@@ -17,6 +17,10 @@
     <ProjectReference Include="..\..\src\AspNet.Security.OAuth.*\*.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Remove="Coinbase\bundle.json" />
+  </ItemGroup>
+
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj]: This isn't needed.

```suggestion
```

File: test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj [hunk 3]
```diff
@@ -31,4 +35,8 @@
     <PackageReference Include="Shouldly" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\..\AspNet.Security.OAuth.Coinbase\AspNet.Security.OAuth.Coinbase.csproj" />
+  </ItemGroup>
+
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj]: This isn't needed.

```suggestion
```

File: test/AspNet.Security.OAuth.Providers.Tests/Coinbase/bundle.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{
+  "$schema": "https://raw.githubusercontent.com/justeat/httpclient-interception/master/src/HttpClientInterception/Bundles/http-request-bundle-schema.json",
+  "items": [
+    {
+      "uri": "http://www.coinbase.com/oauth/token",
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Coinbase/bundle.json]:
```suggestion
      "uri": "https://www.coinbase.com/oauth/token",
```

**Comment by levimatheri** [comment]
@martincostello Thank you for reviewing! I've updated the branch with the changes. Thanks.

**Review by martincostello — CHANGES_REQUESTED** [review]
The whole `AspNet.Security.OAuth.Coinbase` directory needs moving into the `src` directory.

File: AspNet.Security.OAuth.Coinbase/CoinbaseAuthenticationExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        public static AuthenticationBuilder AddCoinbase(
+            [NotNull] this AuthenticationBuilder builder,
+            [NotNull] string scheme,
+            [NotNull] Action<CoinbaseAuthenticationOptions> configuration)
+        {
+            return builder.AddCoinbase(scheme, CoinbaseAuthenticationDefaults.DisplayName, configuration);
+        }
+
+        /// <summary>
+        /// Adds <see cref="DropboxAuthenticationHandler"/> to the specified
```
Comment (martincostello) [inline AspNet.Security.OAuth.Coinbase/CoinbaseAuthenticationExtensions.cs]:
```suggestion
        /// Adds <see cref="CoinbaseAuthenticationHandler"/> to the specified
```

File: test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj [hunk 6]
```diff
@@ -30,5 +30,5 @@
     <PackageReference Include="Moq" />
     <PackageReference Include="Shouldly" />
   </ItemGroup>
-
+
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/AspNet.Security.OAuth.Providers.Tests.csproj]:
```suggestion

```

**Comment by levimatheri** [comment]
@martincostello Sorry about that. Moved to `src\` directory. Thanks.

File: test/AspNet.Security.OAuth.Providers.Tests/Coinbase/CoinbaseTests.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        public override string DefaultScheme => CoinbaseAuthenticationDefaults.AuthenticationScheme;
+
+        protected internal override void RegisterAuthentication(AuthenticationBuilder builder)
+        {
+            builder.AddCoinbase(options => ConfigureDefaults(builder, options));
+        }
+
+        [Theory]
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Coinbase/CoinbaseTests.cs:30]: Just noticed there's no test case for the email claim.

**Review by martincostello — COMMENTED** [review]
One minor change and then this looks good to go.

**Review by martincostello — APPROVED** [review]

File: test/AspNet.Security.OAuth.Providers.Tests/Coinbase/CoinbaseTests.cs (hunk 7 above)
Comment (martincostello) [reply test/AspNet.Security.OAuth.Providers.Tests/Coinbase/CoinbaseTests.cs:30]: I'll add the test for this myself post-merge so I can get the next release shipped.

**Comment by levimatheri** [comment]
@martincostello Thank you!

**Comment by levimatheri** [comment]
@martincostello The request to Coinbase user profile needs to be a GET request. Sorry. Can I push a change to fix this?

**Pull request merged**